### ISSUE TITLE
[ORM] Automated setting of parent object in has_many <-> has_one relationship doesn't respect polymorphic has_one

### ISSUE THREAD

**Issue by HARVS1789UK** [silverstripe/silverstripe-framework#8342]
## Affected Version
`3.7.1`

## Description
In `GridFieldDetailForm_ItemRequest::ItemEditForm()`, in contexts where the GridField sits under the parent `has_one` objects CMS interface (i.e. `Member` has_many `Cars` and `Car` has one `Member` - Security > Member > John Smith > Cars` GridField) the objects foreign key and ID are identified and set automatically, then the corresponding CMS field is made read-only.

This is done around lines 389 - 408 in  `GridFieldDetailForm_ItemRequest::ItemEditForm()`:

```
		// If we are creating a new record in a has-many list, then
		// pre-populate the record's foreign key.
		if($list instanceof HasManyList && !$this->record->isInDB()) {
			$key = $list->getForeignKey();
			$id = $list->getForeignID();
			$this->record->$key = $id;
		}

		$fields = $this->component->getFields();
		if(!$fields) $fields = $this->record->getCMSFields();

		// If we are creating a new record in a has-many list, then
		// Disable the form field as it has no effect.
		if($list instanceof HasManyList) {
			$key = $list->getForeignKey();

			if($field = $fields->dataFieldByName($key)) {
				$fields->makeFieldReadonly($field);
			}
		}
```

This doesn't seem to work correctly when the `has_one` on the other side of the relationship is polymorphic. In this context, the foreign ID field will be set (e.g. `ParentID`) but nothing is done to set the foreign class (e.g. `ParentClass`)?

Am I correct in thinking this is a bug (in which case I can provide more details and examples) or am I missing something?

**Comment by tractorcow** [comment]
Yep, we should have an additional condition for `instanceof PolymorphicHasManyList` right?

**Comment by HARVS1789UK** [comment]
@tractorcow ,

Yes I think so, the below modifications seem to solve the issue for me. They follow the same basis as the current code (automatically setting the relationship fields and making them read only, but they consider polymorphic relationships as well):

```
		// If we are creating a new record in a has-many list, then
		// pre-populate the record's foreign key.
		if ($list instanceof PolymorphicHasManyList && !$this->record->isInDB()) {	
			$idKey = $list->getForeignKey();
			$classKey = substr($idKey, 0, -2) . "Class";
			$class = $list->getForeignClass();
			$id = $list->getForeignID();
			$this->record->$classKey = $class;
			$this->record->$idKey = $id;
		} else if ($list instanceof HasManyList && !$this->record->isInDB()) {
			$key = $list->getForeignKey();
			$id = $list->getForeignID();
			$this->record->$key = $id;
		}

		$fields = $this->component->getFields();
		if(!$fields) $fields = $this->record->getCMSFields();

		// If we are creating a new record in a has-many list, then
		// Disable the form field as it has no effect.

		if ($list instanceof PolymorphicHasManyList) {
			$idKey = $list->getForeignKey();
			$classKey = substr($idKey, 0, -2) . "Class";
			if($classField = $fields->dataFieldByName($classKey)) {
				$fields->makeFieldReadonly($classField);
			}
			if($idField = $fields->dataFieldByName($idKey)) {
				$fields->makeFieldReadonly($idField);
			}
		} else if($list instanceof HasManyList) {
			$key = $list->getForeignKey();
			if($field = $fields->dataFieldByName($key)) {
				$fields->makeFieldReadonly($field);
			}
		}
```

What do you think? Is there a better was of getting the separate foreign key fields instead of doing string manipulation like that?

I could submit a PR for this if you are happy  with it, I just need to know what branch you need me to make it against, `3`, `3.7`, more or less specific?

Cheers,

Joe

**Comment by maxime-rainville** [comment]
@HARVS1789UK Unfortunately, [SilverStripe 3 has entered limited support](https://www.silverstripe.org/blog/update-on-silverstripe-5-x/), which means we'll only release security fixes for it from now on.

If this still affects affects SilverStripe 4, feel free to do a PR against the 4 branch if you want.

**Comment by tractorcow** [comment]
I actually had a similar problem earlier; I think instead of simply setting just the parent id / class, I think the component really needs the ability to specify a default map of field values to assign on construction. I had custom built a component with a `setInitialFields` setter.

**Comment by Cheddam** [comment]
Resolved in #10297, which will ship in `4.12.0`.
